File: frontend/src/app/domain/models/card-invoice.ts
/**
 * One billing cycle ("fatura") of a `credit_card` account. Entirely
 * derived on the backend from the card's `closingDay`/`dueDay` and its
 * ledger - see app/services/card_invoices.py. There is deliberately no
 * cycle math on the frontend: this is the one source of truth, the
 * frontend only renders what it returns.
 *
 * Money fields are decimal strings in the card's own currency, like every
 * other monetary value in the app.
 */
export type CardInvoiceStatus = 'open' | 'closed' | 'overdue' | 'paid' | 'projected';

export interface CardInvoice {
  /** Identifies the cycle - the date its statement closed (or will close). */
  closeDate: string;
  /** First `dueDay` strictly after `closeDate`; when the bill must be paid. */
  dueDate: string;
  /** Inclusive date range of charges in this cycle: (period_start .. period_end). */
  periodStart: string;
  periodEnd: string;
  currency: string;
  /** What the cycle's charges add up to - positive means owed. */
  total: string;
  /** Sum of payments settled against this cycle. */
  paid: string;
  /** `total - paid`; what is still owed. */
  remaining: string;
  status: CardInvoiceStatus;
}

/**
 * Body for paying an invoice. Every field optional: source defaults to the
 * card's `paymentAccountId`, date to today, amount to `remaining`.
 */
export interface CardInvoicePayment {
  accountId?: string;
  date?: string;
  amount?: string;
  description?: string;
}

File: backend/app/schemas/card_invoice.py
"""Credit-card invoice DTOs. An invoice is derived, never stored - see
app/services/card_invoices.py. Money fields are wire-serialized as strings,
like every other monetary DTO.
"""

from datetime import date as date_type
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.schemas.common import serialize_decimal

CardInvoiceStatus = Literal["open", "closed", "overdue", "paid", "projected"]


class CardInvoiceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    close_date: date_type
    due_date: date_type
    period_start: date_type
    period_end: date_type
    currency: str
    total: Decimal
    paid: Decimal
    remaining: Decimal
    status: CardInvoiceStatus

    @field_serializer("total", "paid", "remaining")
    def _serialize_money(self, value: Decimal) -> str | None:
        return serialize_decimal(value)


class CardInvoicePaymentCreate(BaseModel):
    """Body for POST /accounts/{id}/invoices/{close_date}/pay. Every field
    is optional: source defaults to the card's `payment_account_id`, date to
    today, amount to the invoice's remaining balance."""

    account_id: UUID | None = None
    date: date_type | None = None
    amount: Decimal | None = Field(default=None, gt=0)
    description: str | None = Field(default=None, min_length=1, max_length=200)
